vertical-pod-autoscaler/pkg/recommender/logic/estimator.go

```diff
@@ -152,9 +158,31 @@ func (e *minResourcesEstimator) GetResourceEstimation(s *model.AggregateContaine
 	newResources := make(model.Resources)
 	for resource, resourceAmount := range originalResources {
 		if resourceAmount < e.minResources[resource] {
+			if resource == "memory" {
+				klog.Warningf("Computed resources for %s were below minimum! Computed %v, minimum is %v.", resource, resourceAmount, e.minResources[resource])
+				logHistogramInformation(s)
+				metrics_quality.ObserveLowerThanMinRecommendation(s.GetUpdateMode(), corev1.ResourceName(resource), e.vpaKey.Namespace+"/"+e.vpaKey.VpaName)
+			}
 			resourceAmount = e.minResources[resource]
 		}
 		newResources[resource] = resourceAmount
 	}
 	return newResources
 }
+
+func logHistogramInformation(s *model.AggregateContainerState) {
+	if s.AggregateCPUUsage == nil {
+		klog.Warning("Aggregate CPU usage has no metric samples, cannot show internal histogram data!")
+		return
+	}
+	if s.AggregateMemoryPeaks == nil {
+		klog.Warning("Aggregate memory usage has no metric samples, cannot show internal histogram data!")
+		return
+	}
+	c, _ := s.SaveToCheckpoint()
+	prettyCheckpoint, err := json.MarshalIndent(c, "", "  ")
+	if err != nil {
+		klog.Errorf("Error during marshalling checkpoint: %s", err)
```

Comment:
Maybe return here? It does not make sense to continue and to log `prettyCheckpoint` if we failed to marshal it.